### PR TITLE
add master design

### PR DESCRIPTION
**Pull request by skydoorkai**

### CODE REVIEW

File: docs/designs/master_redesign.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+### Elastic Pod Manager
+The elastic pod manager is responsible for:
+
+1. Create PS/worker pods;
+2. Monitor pod events, keep track all pods status;
+3. Relaunch failed worker pods when needed.
+
+### Task Manager
+
+The task manager is responsible for creating tasks. Workers will ask the task manager for new tasks and report task completion status to the task manager when needed. Optionally, the task mamager supports task fault-tolerance by recovering un-completed tasks from failed workers and re-assigning to other workers.
```
Comment (workingloong) [inline docs/designs/master_redesign.md]: The task manager is responsible for creating tasks and dispatching tasks to workers.

Comment (workingloong) [inline docs/designs/master_redesign.md]: Why does not the worker always report task completion status?

File: docs/designs/master_redesign.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+For training jobs, the task manager supports dynamic sharding and partitions all training data into shards. Each task is a data shard.
```
Comment (workingloong) [inline docs/designs/master_redesign.md]: Each task contains a data shard.

File: docs/designs/master_redesign.md (hunk 1 above)
Comment (skydoorkai) [reply docs/designs/master_redesign.md]: If the task manager does not support task fault-tolerance, no need to report task completion.

File: docs/designs/master_redesign.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+The task manager is responsible for creating tasks and dispatching tasks to workers. Workers will ask the task manager for new tasks and report task completion status to the task manager when needed. Optionally, the task mamager supports task fault-tolerance by recovering un-completed tasks from failed workers and re-assigning to other workers.
+
+For training jobs, the task manager supports dynamic sharding and partitions all training data into shards. Each task contains a data shard.
+
+### Rendezvous Server
+
+The rendezvous server supports for elastic AllReduce. After detecting worker pod number change, it creates a new worker host plan so that worker pods can re-initialize Horovod.
+
+### ElasticDL Training/Evaluation Service
+All other current master functions except the above three components. ElasticDL training/evaluation service supports for ElasticDL native training. It includes evaluation service, supporting for model saving and checkpoint, etc.
```
Comment (brightcoder01) [inline docs/designs/master_redesign.md]: native training -> native training loop.
It includes evaluation service, supporting for model saving and checkpoint, etc. => Why evaluation service is responsible for exporting the model and save checkpoints?

File: docs/designs/master_redesign.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+### Rendezvous Server
+
+The rendezvous server supports for elastic AllReduce. After detecting worker pod number change, it creates a new worker host plan so that worker pods can re-initialize Horovod.
+
+### ElasticDL Training/Evaluation Service
+All other current master functions except the above three components. ElasticDL training/evaluation service supports for ElasticDL native training. It includes evaluation service, supporting for model saving and checkpoint, etc. 
+
+
+Thus, we can launch the master in seven component combinations:
```
Comment (brightcoder01) [inline docs/designs/master_redesign.md:50]: How do we specify which components we should launch? Use a command arguments?

File: docs/designs/master_redesign.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+ rendezvous_server = create_rendezvous_server_if_needed(args, pod_manager)
+ 
+ # init ElasticDL training/evaluation service if args.need_training_service
+ training_service = init_training_service_if_needed(args)
+ 
+ # create master rpc service
+ rpc_server = create_rpc_service()
+
+
+Prepare
```
Comment (workingloong) [inline docs/designs/master_redesign.md:134]: The structure is similar to the initialization:
https://github.com/sql-machine-learning/elasticdl/blob/d2015d1a68f9b28c6a412dd75a1419d2e7b1ee03/elasticdl/python/master/servicer.py#L32-L40.

We should take the master and `MasterServicer` into consideration to design the master.

File: docs/designs/master_redesign.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+Run
+  loop until no more tasks, or all workers completed.
+
+```
+
+Note that k8s client with watch is included in the pod manager, which is using an event callback defined inside the pod manager. Thus, both the task manager with fault-tolerance and the rendezvous server need to enable its callback functions in that event callback.
+
+### Implementation Steps
+
+1. Create the above framework, but put everything in the current master version into training_service. We will remove most components from it after implementing them as modular components.
```
Comment (brightcoder01) [inline docs/designs/master_redesign.md]: remove most component -> move most component out of it

File: docs/designs/master_redesign.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
+# Master Design
+
+This document deseribes an ElasticDL master design.
```
Comment (brightcoder01) [inline docs/designs/master_redesign.md]: This document deseribes an ElasticDL master design. -> This document is a new design of the ElasticDL master.

File: docs/designs/master_redesign.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+## Motivation
+
+We plan to provide elastic training feature to customized AllReduce training codes and other deep learning frameworks. Currently the master is designed for ElasticDL's own training framework. This requires us to re-design the master.
```
Comment (brightcoder01) [inline docs/designs/master_redesign.md]: customized AllReduce training codes -> custom training loop for AllReduce

File: docs/designs/master_redesign.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+After this re-design, the master consists of a few modular components and only launches those required components when providing elastic training feature to others.
```
Comment (brightcoder01) [inline docs/designs/master_redesign.md]: launches various required components while providing elastic training feature in different scenarios.

File: docs/designs/master_redesign.md (hunk 5 above)
Comment (skydoorkai) [reply docs/designs/master_redesign.md:134]: We will leave master rpc service into another issue.

File: docs/designs/master_redesign.md (hunk 3 above)
Comment (skydoorkai) [reply docs/designs/master_redesign.md]: valuation service, supporting for model saving and checkpoint. There are 3 functions inside ElasticDL Training/Evaluation Service, not the first one includes the following two.

File: docs/designs/master_redesign.md (hunk 4 above)
Comment (skydoorkai) [reply docs/designs/master_redesign.md:50]: As mentioned in Implementation, we will add more arguments to specify which components are needed.

**Comment by workingloong** [comment]
You should fix the markdownlint where the line length is less than 80.

**Review by workingloong — APPROVED** [review]
LGTM

**Pull request merged**